src/mbgl/text/glyph_set.cpp

```diff
@@ -99,70 +99,140 @@ void justifyLine(std::vector<PositionedGlyph>& positionedGlyphs,
     }
 }
 
-float GlyphSet::determineIdeographicLineWidth(const std::u16string& logicalInput,
+float GlyphSet::determineAverageLineWidth(const std::u16string& logicalInput,
                                               const float spacing,
                                               float maxWidth) const {
     float totalWidth = 0;
 
-    // totalWidth doesn't include the last character for magical tuning reasons. This makes the
-    // algorithm a little
-    //  more agressive about trying to fit the text into fewer lines, taking advantage of the
-    //  tolerance for going a little
-    //  over maxWidth
-    for (uint32_t i = 0; i < logicalInput.size() - 1; i++) {
-        auto it = sdfs.find(logicalInput[i]);
-        if (it != sdfs.end())
+    for (char16_t chr : logicalInput) {
+        auto it = sdfs.find(chr);
+        if (it != sdfs.end()) {
             totalWidth += it->second.metrics.advance + spacing;
+        }
     }
 
     int32_t lineCount = std::fmax(1, std::ceil(totalWidth / maxWidth));
     return totalWidth / lineCount;
 }
+    
+float calculateBadness(const float lineWidth, const float targetWidth, const float penalty, const bool isLastBreak) {
+    const float raggedness = std::pow(lineWidth - targetWidth, 2);
+    if (isLastBreak && lineWidth < targetWidth) {
+        // Be more tolerant of short final lines
+        return std::fmax(0, raggedness - 150);
+    }
+    if (penalty < 0) {
+        return raggedness - std::pow(penalty, 2);
+    }
+    return raggedness + std::pow(penalty, 2);
+}
+    
+float calculatePenalty(char16_t codePoint, char16_t previousCodePoint) {
+    float penalty = 0;
+    // Force break on newline
+    if (codePoint == 0x0a) {
+        penalty -= 10000;
+    }
+    // Penalize open parenthesis at end of line
+    if (previousCodePoint && (previousCodePoint == 0x28 || previousCodePoint == 0xff08)) {
+        penalty += 50;
+    }
+    // Penalize close parenthesis at beginning of line
+    if (codePoint == 0x29 || codePoint == 0xff09) {
+        penalty += 50;
+    }
+    return penalty;
+}
+    
+struct PotentialBreak {
+    const int32_t index;
+    const float x;
+    const PotentialBreak* priorBreak;
+    const float badness;
+};
+    
+struct LineBreakEdge {
+    const PotentialBreak* priorBreak;
+    const float badness;
+};
+    
+PotentialBreak evaluateBreak(const int32_t breakIndex, const float breakX, const float targetWidth, const std::list<PotentialBreak>& potentialBreaks, const float penalty, const bool isLastBreak) {
+    // We could skip evaluating breaks where the line length (breakX - priorBreak.x) > maxWidth
+    //  ...but in fact we allow lines longer than maxWidth (if there's no break points)
+    //  ...and when targetWidth and maxWidth are close, strictly enforcing maxWidth can give
+    //     more lopsided results.
+    
+    std::list<LineBreakEdge> edges;
```

Comment:
Ah ok. I think it's better to avoid `shared_ptr`, either by using a container that doesn't invalidate references on insert or by precalculating the number of potential breakpoints and preallocating the vector.